### ISSUE TITLE
False positive with `redundant_type_annotations` lint

### ISSUE THREAD

**Issue by bpandreotti** [rust-lang/rust-clippy#12212]
### Summary

The `redundant_type_annotations` lint is returning a false positive when used with byte string literals (i.e. `b""`).

### Lint Name

redundant_type_annotations

### Reproducer

When running on the following snippet of code, clippy suggests that the type annotation for `a` is not necessary:

```rust
#![warn(clippy::redundant_type_annotations)]

use std::io::BufRead;

fn foo<T: BufRead>(_: T) {}

fn main() {
    let a: &[u8] = b"hello";
    foo(a);
}
```

However, if you remove the type annotation, `a`'s type is inferred to be `&[u8; 5]` instead of `&[u8]`. And, since `&[u8; 5]` does not implement `BufRead`, the call to `foo` causes a compile error. 

### Version

```text
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: x86_64-unknown-linux-gnu
release: 1.75.0
LLVM version: 17.0.6
```
